src/tools/cephfs_mirror/PeerReplayer.cc

```diff
@@ -652,15 +660,130 @@ int PeerReplayer::remote_mkdir(const std::string &epath, const struct ceph_statx
+  return file_size_threshold * max_concurrent_large_file +
+         std::min(file_size, file_size_threshold) *
+             (num_threads - max_concurrent_large_file);
+}
+
+void PeerReplayer::FileTransferHandlerThreadPool::run() {
+  while (true) {
+    std::function<void(bool)> task;
+    int64_t token;
+    {
+      std::unique_lock<std::mutex> lock(pmtx);
+      pick_task.wait(
+          lock, [this] { return (stop_flag.load() || !task_queue.empty()); });
+      if (stop_flag.load()) {
+        return;
+      }
+      std::tie(task, token) = std::move(task_queue.front());
+      task_queue.pop();
+      give_task.notify_one();
+      exec_task.wait(lock, [&token, this] {
+        return (stop_flag.load() || current_token_sum + token <= max_token_sum);
+      });
+      if (stop_flag.load()) {
+        return;
+      }
+      current_token_sum += token;
+    }
+    task(false);
+    {
+      std::scoped_lock<std::mutex> lock(pmtx);
+      current_token_sum -= token;
+      exec_task.notify_one();
+    }
+  }
+}
+
+template <class F, class... Args>
+void PeerReplayer::FileTransferHandlerThreadPool::enqueue(
+    std::atomic<int> &counter, std::condition_variable &dir_cv,
+    uint64_t file_size, F &&f, Args &&...args) {
+  {
+    std::unique_lock<std::mutex> lock(pmtx);
+    uint64_t token = get_token(file_size);
+    give_task.wait(lock, [this] {
+      return (stop_flag.load() || task_queue.size() < task_queue_limit);
+    });
+    if (stop_flag.load()) {
+      return;
+    }
+    counter++;
+    task_queue.emplace(
+        [f = std::forward<F>(f), ... args = std::forward<Args>(args), &counter,
+         &dir_cv](bool dont_execute) mutable {
+          if (!dont_execute) {
+            f(std::move(args)...);
+          }
+          --counter;
+          if (counter.load() == 0) {
+            dir_cv.notify_one();
+          }
+        },
+        token);
+  }
+  pick_task.notify_one();
+}
+
 #define NR_IOVECS 8 // # iovecs
 #define IOVEC_SIZE (8 * 1024 * 1024) // buffer size for each iovec
-int PeerReplayer::copy_to_remote(const std::string &dir_root,  const std::string &epath,
-                                 const struct ceph_statx &stx, const FHandles &fh) {
+int PeerReplayer::copy_to_remote(const std::string &dir_root,
+                                 const std::string &epath,
+                                 const struct ceph_statx &stx,
+                                 const FHandles &fh,
+                                 std::atomic<int> &rem_file_transfer,
+                                 std::condition_variable &cv,
+                                 std::shared_ptr<std::atomic<bool>> &canceled,
+                                 std::shared_ptr<std::atomic<bool>> &failed) {
   dout(10) << ": dir_root=" << dir_root << ", epath=" << epath << dendl;
   int l_fd;
   int r_fd;
-  void *ptr;
-  struct iovec iov[NR_IOVECS];
 
   int r = ceph_openat(m_local_mount, fh.c_fd, epath.c_str(), O_RDONLY | O_NOFOLLOW, 0);
```

Comment:
Why not opening source/target files inside transfer()?
Not doing that might result in tons of file descriptors being opened while pending transfers in the queue...